compiler/rustc_span/src/symbol.rs

```diff
@@ -1710,46 +1710,68 @@ pub(crate) struct Interner(Lock<InternerInner>);
 // This type is private to prevent accidentally constructing more than one `Interner` on the same
 // thread, which makes it easy to mixup `Symbol`s between `Interner`s.
 #[derive(Default)]
-struct InternerInner {
+struct InternerShard {
     arena: DroplessArena,
     names: FxHashMap<&'static str, Symbol>,
     strings: Vec<&'static str>,
 }
 
 impl Interner {
-    fn prefill(init: &[&'static str]) -> Self {
-        Interner(Lock::new(InternerInner {
-            strings: init.into(),
-            names: init.iter().copied().zip((0..).map(Symbol::new)).collect(),
-            ..Default::default()
+    #[cfg(test)]
+    fn empty_for_test() -> Self {
+        Self([(); 256].map(|()| Lock::new(InternerShard::default())))
+    }
+
+    fn prefill(init: &[&[&'static str]; 256]) -> Self {
+        let mut i = 0;
+        Interner(init.map(|init| {
+            let shard = Lock::new(InternerShard {
+                strings: init.into(),
+                names: init
+                    .iter()
+                    .copied()
+                    .zip((0..).map(|idx| Symbol::new(idx << 8 | i)))
+                    .collect(),
+                ..Default::default()
+            });
+            i += 1;
+            shard
         }))
     }
 
     #[inline]
     fn intern(&self, string: &str) -> Symbol {
-        let mut inner = self.0.lock();
-        if let Some(&name) = inner.names.get(string) {
-            return name;
-        }
+        // Warning: hasher has to be kept in sync with rustc_macros::symbols to ensure that all
+        // pre-filled symbols are assigned the correct shard.
+        let mut state = FxHasher::default();
+        string.hash(&mut state);
+        let hash = state.finish();
+
+        let mut shard = self.0[(hash & 0xff) as usize].lock();
```

Comment:
Shouldn't you `hash >> 8` or similar here? Otherwise the hashmap is going to be colliding constantly since the low bits are used to index into the backing storage, IIRC.

Reply:
The symbol integer is created as `index_within_shard << 8 | shard_num`.

Reply:
Maybe I'm missing something, but `shard.names.raw_entry_mut().from_key_hashed_nocheck(hash, string)` looks like it's using just the plain hash, not shifting based on the sharding. I think the low bits there will always be the same, which seems like a problem.